controller_server:
  ros__parameters:
    controller_frequency: 5.0
    costmap_update_timeout: 0.30
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.001 # 0.5 for differential drive robots that have no y velocity
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.3
    odom_topic: "odometry"
    progress_checker_plugins: ["progress_checker"]
    goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
    controller_plugins: ["FollowPath"]
    use_realtime_priority: false

    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 10.0

    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.35
      yaw_goal_tolerance: 0.50

    FollowPath:
      plugin: "mappi::MPPIcROS"

      GeneralSettings:
        global_frame: ""
        local_frame: ""
        transform_tolerance: 0.5
        num_iterations: 1
        batch_size: 1000
        time_steps: 100
        num_retry: 4
        control_offset: 1
        goal_tolerance: 0.5
        plan_shift: 1.0
        use_splines: false
        model_dt: 0.05
        temperature: 0.3
        gamma: 0.05

      Constraints:
        max_vx: 0.75
        min_vx: -0.75
        max_vy: 0.5
        min_vy: -0.5
        max_wz: 0.5
        min_wz: -0.5

      MotionModel: "BicycleKin"

      Ackermann:
        min_radius: 3.0

      BicycleKin:
        length: 0.5
        max_steer: 0.3

      NoiseGeneration:
        std_vx: 0.3
        std_vy: 0.3
        std_wz: 0.1

      Critics:
        Goal:
          active: true
          power: 1
          weight: 1.0
          threshold: 1.5
        GoalAngle:
          active: true
          power: 1
          weight: 1.0
          threshold: 1.5
        PathDist:
          active: false
          power: 2
          weight: 5.0
          threshold: 1.0
          stride: 2
        Twirling:
          active: false
          power: 1
          weight: 10.0
        Forward:
          active: false
          power: 1
          weight: 10.0
        PathFollow:
          active: true
          power: 1
          weight: 10.0
          threshold: 1.0
          offset_from_furthest: 3
        PathAngle:
          active: false
          power: 1
          weight: 25.0
          threshold: 1.0
          offset_from_furthest: 3
          angle_threshold: 90.0
        Obstacles:
          active: true
          power: 1
          weight: 50.0
          threshold: 1.0
          repulsive_weight: 0.0
          collision_cost: 100000.0
          collision_margin_dist: 0.2

      Visualization:
        active: true
        batch_stride: 100
        time_stride: 3
        default_z: 0.2
        scale: 0.05


local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 5.0
      global_frame: odom
      robot_base_frame: base_link
      rolling_window: true
      width: 15
      height: 15
      resolution: 0.05
      robot_radius: 0.15  
      plugins: ["obstacle_layer", "denoise_layer", "inflation_layer"]
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        scan:
          topic: /ona2/scan
          raytrace_min_range: 0.20
          obstacle_min_range: 0.20
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "LaserScan"
      denoise_layer:
        plugin: "nav2_costmap_2d::DenoiseLayer"
        enabled: true
        minimal_group_size: 2
        group_connectivity_type: 8
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 2.58
        inflation_radius: 1.75
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: True
      always_send_full_costmap: True

global_costmap:
  global_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 5.0
      global_frame: map
      robot_base_frame: base_link
      robot_radius: 0.15  
      resolution: 0.05
      track_unknown_space: true
      plugins: ["static_layer", "obstacle_layer", "inflation_layer"]
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: True
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        scan:
          topic: /ona2/scan
          raytrace_min_range: 0.20
          obstacle_min_range: 0.20
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "LaserScan"
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 2.58
        inflation_radius: 1.75
      always_send_full_costmap: True

map_saver:
  ros__parameters:
    save_map_timeout: 5.0
    free_thresh_default: 0.25
    occupied_thresh_default: 0.65
    map_subscribe_transient_local: True

planner_server:
  ros__parameters:
    planner_plugins: ["GridBased"]

    GridBased:
      plugin: "nav2_smac_planner::SmacPlannerHybrid" 
      downsample_costmap: false           # whether or not to downsample the map
      downsampling_factor: 1              # multiplier for the resolution of the costmap layer (e.g. 2 on a 5cm costmap would be 10cm)
      tolerance: 0.25                     # dist-to-goal heuristic cost (distance) for valid tolerance endpoints if exact goal cannot be found.
      allow_unknown: true                 # allow traveling in unknown space
      max_iterations: 1000000             # maximum total iterations to search for before failing (in case unreachable), set to -1 to disable
      max_on_approach_iterations: 1000    # Maximum number of iterations after within tolerances to continue to try to find exact solution
      max_planning_time: 5.0              # max time in s for planner to plan, smooth
      motion_model_for_search: "DUBIN"    # Hybrid-A* Dubin, Redds-Shepp
      angle_quantization_bins: 72         # Number of angle bins for search
      analytic_expansion_ratio: 3.5       # The ratio to attempt analytic expansions during search for final approach.
      analytic_expansion_max_length: 3.0  # For Hybrid/Lattice nodes: The maximum length of the analytic expansion to be considered valid to prevent unsafe shortcutting
      analytic_expansion_max_cost: 200.0  # The maximum single cost for any part of an analytic expansion to contain and be valid, except when necessary on approach to goal
      analytic_expansion_max_cost_override: false  #  Whether or not to override the maximum cost setting if within critical distance to goal (ie probably required)
      minimum_turning_radius: 0.40        # minimum turning radius in m of path / vehicle
      reverse_penalty: 2.0                # Penalty to apply if motion is reversing, must be => 1
      change_penalty: 0.0                 # Penalty to apply if motion is changing directions (L to R), must be >= 0
      non_straight_penalty: 1.2           # Penalty to apply if motion is non-straight, must be => 1
      cost_penalty: 2.0                   # Penalty to apply to higher cost areas when adding into the obstacle map dynamic programming distance expansion heuristic. This drives the robot more towards the center of passages. A value between 1.3 - 3.5 is reasonable.
      retrospective_penalty: 0.015
      lookup_table_size: 20.0             # Size of the dubin/reeds-sheep distance window to cache, in meters.
      cache_obstacle_heuristic: false     # Cache the obstacle map dynamic programming distance expansion heuristic between subsequent replannings of the same goal location. Dramatically speeds up replanning performance (40x) if costmap is largely static.
      debug_visualizations: false         # For Hybrid nodes: Whether to publish expansions on the /expansions topic as an array of poses (the orientation has no meaning) and the path's footprints on the /planned_footprints topic. WARNING: heavy to compute and to display, for debug only as it degrades the performance.
      use_quadratic_cost_penalty: False
      downsample_obstacle_heuristic: True
      allow_primitive_interpolation: False
      coarse_search_resolution: 4         # Number of bins to skip when doing a coarse search for the path. Only used for all_direction goal heading mode.
      goal_heading_mode: "DEFAULT"        # DEFAULT, BIDIRECTIONAL, ALL_DIRECTION
      smooth_path: True                   # If true, does a simple and quick smoothing post-processing to the path

      smoother:
        max_iterations: 1000
        w_smooth: 0.3
        w_data: 0.2
        tolerance: 1.0e-10
        do_refinement: true
        refinement_num: 2